### PR TITLE
[Work and discussion in progress] Add `Unguarded` syntax for (co)fixpoints and `Assumed Positive` for (co)inductive types.

### PR DESCRIPTION
**Pull request by SimonBoulier**
Disable guard check for (co)fixpoints when defined with `Ungarded` keyword.
We can now write:
```
Unguarded Fixpoint f (n : nat) : False
  := f n.

Unguarded Fixpoint ackermann (m n : nat) {struct m} : nat.
  destruct m. exact (S n).
  destruct n. exact (ackermann m 1).
  exact (ackermann m (ackermann (S m) n)).
Defined.

Compute (ackermann 3 4).

Print Assumptions ackermann.
(* ackermann is assumed to be guarded *)

Unguarded CoFixpoint fer (s : Stream nat) : Stream nat.
Proof.
  exact (fer s).
Defined.
```

**Kind:** feature.

- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in CHANGES.

I know that this PR is far from being perfect, any comment welcome.

### CODE REVIEW

File: CHANGES
Comment (Zimmi48) [inline CHANGES]: typo: spurious s

File: doc/sphinx/language/gallina-specification-language.rst
Comment (Zimmi48) [inline doc/sphinx/language/gallina-specification-language.rst]: You should also add this on the CoFixpoint line below.

File: doc/sphinx/language/gallina-specification-language.rst
Comment (Zimmi48) [inline doc/sphinx/language/gallina-specification-language.rst]: Note that this is already fixed in #7537 and the best fix goes further than this.

File: doc/sphinx/language/gallina-specification-language.rst
Comment (Zimmi48) [inline doc/sphinx/language/gallina-specification-language.rst]: This documentation is not done in the recommended style, the documentation must come after the variant grammar and must be indented. Since this chapter is heavily reindented by #7537 anyways, I suggest you rebase on top or wait for it to get merged. In any case, you will have so many conflicts in this file that the best way to solve them will be to copy / paste this part.

**Review by Zimmi48 — COMMENTED** [review]
Since we can also do fixed points using `Theorem`, `Lemma`, etc, would it make sense to also add this new syntax for these commands?

File: doc/sphinx/language/gallina-specification-language.rst
Comment (SimonBoulier) [reply doc/sphinx/language/gallina-specification-language.rst]: Ok thanks, I will do this after merge then.

File: vernac/comFixpoint.ml
Comment (SimonBoulier) [reply vernac/comFixpoint.ml]: I named guarded, but do you suggest I also name the other arguments?

Comment (ejgallego) [reply vernac/comFixpoint.ml]: Maybe that would be a good idea, what do other devs think?

**Comment by Zimmi48** [comment]
> Since we can also do fixed points using Theorem, Lemma, etc, would it make sense to also add this new syntax for these commands?

I withdraw this because I'm actually doubtful that doing fixed points using `Theorem` is a good idea anyways.

**Comment by ppedrot** [comment]
***Please. Stop. Inflating. The. Definition. Modifier. Syntax. Again.***

There have been a shitload of arguments against definition modifiers, and it all boils down that it doesn't scale. I am too tired to repeat them here and I think there are pointers in #79.  I am fundamentally against this PR.

**Review by ppedrot — CHANGES_REQUESTED** [review]
To expand on my above comment, I am even surprised that people that were (I think) at the heated Coq workgroup where we discussed the issue comment on that PR as if it was OK in principle, and that they did not try to deter @SimonBoulier from implementing it at the CIW... Is that the famous "History repeats twice, first as a tragedy, second as a farce" effect?

**Comment by Zimmi48** [comment]
@SimonBoulier The documentation PR has been merged so you might want to rebase now.

@ppedrot I completely disagree with your argument to reject this change. This argument has been stupidly used to delay #79 for three years now. The problem you are pointing at is a well known syntax problem and must be addressed independently without preventing people to work on new features in the meantime. Attributes are now on their way thanks to @vbgl and will probably be finished for 8.9 so it means that the current syntax will probably never make it into a released version as-is. But even if it wasn't the case, that wouldn't be a good reason to delay the availability of this feature.

**Comment by ejgallego** [comment]
I like this feature and its current syntax. We are taking attributes as a kind of "silver bullet" , but the more I think about them the less convinced I am they are going to bring significant benefits.

Attributes are _hard_ to handle, and worse, they tend to override all kind of static checks due to their "generic" nature. So IMVHO we shouldn't delay this PR just for that.

**Comment by maximedenes** [comment]
> So IMVHO we shouldn't delay this PR just for that.

I tend to agree with that part.

**Comment by ppedrot** [comment]
> The problem you are pointing at is a well known syntax problem and must be addressed independently without preventing people to work on new features in the meantime. 

Wrong. Syntax is non-local by nature, so this can affect people that are unrelated to this change. Plus in the current implementation it is a hell to get it right. There is no way to compute the syntax conflict because CAMLP5, but we would be surprised if this were possible. This is the kind of area where conservativity should be of utmost importance, because it will be hard to backtrack on that.

But like, I don't understand why it is taking THREE YEARS. Like I can implement a syntax that scales properly in literally half a day. Should I do that or what?

**Comment by ppedrot** [comment]
And also,

> But even if it wasn't the case, that wouldn't be a good reason to delay the availability of this feature.

We've already discussed that, and the answer is: write a goddamn plugin.

**Comment by Zimmi48** [comment]
> But like, I don't understand why it is taking THREE YEARS. Like I can implement a syntax that scales properly in literally half a day. Should I do that or what?

Seriously? And you didn't do it?
Anyways, now your help on this is not needed anymore since it is well underway thanks to @vbgl.

**Comment by ppedrot** [comment]
> Seriously? And you didn't do it?

This ended up in the sinkhole of workgroups and PR bikeshedding, as all syntactic considerations do. But yes, luckily @vbgl has started doing it with a wide approval, although I am still curious to see the moment where we start discussing on the actual syntax.

**Comment by ejgallego** [comment]
IMO "Write a plugin" is hardly an answer to most problems, but rather a good way to add some maintenance burden and testing overhead. This seems to me like a well-delimited functionality that applies to two core vernaculars.

**Review by ejgallego — CHANGES_REQUESTED** [review]
I am convinced now that this property is local to the vernaculars and thus doesn't belong in an attribute.

File: vernac/vernacinterp.mli
Comment (ejgallego) [reply vernac/vernacinterp.mli]: I mean this should not be an attribute, but an entry in the corresponding `vernac_expr` constructor.

**Comment by Zimmi48** [comment]
It seems Emilio that you have a diverging opinion of what should be an attribute than what was in the CEP and what is being implemented. IMO the great benefit of attributes is that they make the parsing phase more tractable by moving some of the complexity / error handling from the parsing file to the interpretation phase.

**Comment by gares** [comment]
> this property is local to the vernaculars

Can you explain what you mean?

**Comment by gares** [comment]
I like the feature, I'm fine with the current "prefix" syntax and the plan to turn it into one of the many tags/attributes. I dislike the word `Unguarded`, it is too obscure. All users understand "termination checking" or "total functions" but "guardedness condition" is, frankly, even hard to pronounce.

One thing is quite fishy to my eyes, please @SimonBoulier answer the following concern.
Fixpoints are not named, eg Fixpoint is just `Definition` + `fix`. I wonder how print assumption can detect the unguarded fixpoint if it is copied in my proof by unifying a goal with the type of a lemma (containing the unguarded fixpoint) and unification has unfolded the fixpoint (i.e. lose his name). Is the `fix` term constructor containing the `tainted` bit (which I doubt given how little files did change) or is the `Definition` containing it?
I'm not 100% sure I can prove False undetected, but maybe @maximedenes can ;-)

**Comment by ejgallego** [comment]
Well indeed maybe "convinced" is far too strong. I tend to have a more "strongly typed" vision of commands, where the type for each vernacular reflects what the command depends on.

I never understood the attributes proposal in a very wide scope. To my point of view attributes are something that are, in principle, data pplicable to most commands, such as location, locality, or preprocessing information. However here we have only two commands that can understand the unguarded bit so I am not sure that should be an attribute. Certainly when introducing the attr record I was not thinking of that case.

It would seem to me that such information would be an attribute if attached to terms, in such a way that would mean "don't check termination" in the checking of that term.

But I dunno, I guess I'm certainly fine if people thinks that we should use a vernac attribute for that, but I fear that we end up with a record which is a bit messy in the sense of having weak "uniformity" properties. Maybe a good solution is indeed to further refine "attributes" as to have more representative classes. For example we could have "definition_attr" for attributes specific to definition-like commands, "scoped_obj" attributes for commands that support scope, etc...

Anyways take my opinion more as an evolving reflection than as a strong statement on the solution to choose.

**Comment by ejgallego** [comment]
I would certainly enjoy github refining the "red" review button as to have a "strong" and "weak" request for changes.

**Comment by gares** [comment]
OK, but I still don't get exactly what your review talks about. Here I see
- a feature (letting one switch off termination checking)
- a syntax in the current style and a proposal (switch to a uniform attribute syntax)
- an implementation

I read @ppedrot and @ejgallego reviews as "No to the syntax". I fully understand this POV, even if I've a different opinion. *but* this PR is more than just some syntax, namely it adds a feature and an provides an implementation. Do you guys agree that the feature is OK? Do you have remarks about the implementation?

**Comment by ejgallego** [comment]
Ive only looked at the connection between the syntax and the implementation so far, but not at the implementation itself.

**Comment by Zimmi48** [comment]
> One thing is quite fishy to my eyes, please @SimonBoulier answer the following concern.
Fixpoints are not named, eg Fixpoint is just Definition + fix. I wonder how Print Assumptions can detect the unguarded fixpoint if it is copied in my proof by unifying a goal with the type of a lemma (containing the unguarded fixpoint) and unification has unfolded the fixpoint (i.e. lose his name). Is the fix term constructor containing the tainted bit (which I doubt given how little files did change) or is the Definition env entry that contains it?

Simon explained that this new feature doesn't go as far as he would have liked because the check disabling only occurs at definition time. So if the fixpoint ended up in a proof it would have to be typechecked again and this would fail.

**Comment by Zimmi48** [comment]
> Certainly when introducing the attr record I was not thinking of that case.

Indeed, it is funny that you were the one that introduced this attr record but had a completely different understanding of what an attribute was compared to most other developers. Remember how long it took to merge this PR? That was in part due to the very high confusion surrounding its purpose. Contrary to the PR that introduced the attr records, the attribute feature has had a CEP which is clear and detailed so I would defer any further design discussion there.

> But I dunno, I guess I'm certainly fine if people thinks that we should use a vernac attribute for that, but I fear that we end up with a record which is a bit messy in the sense of having weak "uniformity" properties. Maybe a good solution is indeed to further refine "attributes" as to have more representative classes. For example we could have "definition_attr" for attributes specific to definition-like commands, "scoped_obj" attributes for commands that support scope, etc...

I would personally push for an attribute record that is extensible, even if it requires some form of dynamic type checking, and I've said so to @vbgl in private.

> It would seem to me that such information would be an attribute if attached to terms, in such a way that would mean "don't check termination" in the checking of that term.

This would be better than the current proposal but would also require a much deeper refactoring that I'm not sure that Simon is ready to undertake.

> I would certainly enjoy github refining the "red" review button as to have a "strong" and "weak" request for changes.

I think you should just dismiss your "request for changes" to transform it into a simple "comment review".

> a syntax in the current style and a proposal (switch to a uniform attribute syntax)

The current proposal uses the attribute system internally but not on the user-side because there is no syntax for attributes yet.

File: library/decl_kinds.ml
Comment (Zimmi48) [inline library/decl_kinds.ml]: I don't like `bool option`. Why don't you create a three-valued type like is done around? Why do you need three values BTW?

File: printing/printer.ml
Comment (Zimmi48) [inline printing/printer.ml]: If this is not affecting any output test, could you add one? I'd like to understand in which situation these messages are seen.

File: stm/vernac_classifier.ml
Comment (Zimmi48) [inline stm/vernac_classifier.ml]: This is not going to scale if we get many attributes. You should just use `_ ->` here.

File: vernac/comFixpoint.mli
Comment (Zimmi48) [inline vernac/comFixpoint.mli]: The type signature doesn't correspond to the implementation (compilation is broken).
Can you make these three first parameters named ones since you are changing the API anyways?
BTW, what is this "Not used so far" comment above?

File: vernac/comFixpoint.mli
Comment (Zimmi48) [inline vernac/comFixpoint.mli]: The removal of the comment is only making things less clear because the semantics of `check_guard` is non-obvious.

File: vernac/g_vernac.ml4
Comment (Zimmi48) [inline vernac/g_vernac.ml4]: Why don't you define this option? It should be really easy to add one. Indeed, this had me confused about the reasons for a three-valued flag.

File: vernac/vernacentries.ml
Comment (Zimmi48) [inline vernac/vernacentries.ml]: @ejgallego Cf. here and the above function about the Polymorphic flag to better understand the attribute system.

**Comment by ejgallego** [comment]
> the attribute feature has had a CEP which is clear and detailed so I would defer any further design discussion there.

I am not sure which CEP you are referring to. If that's https://github.com/aspiwack/ceps/blob/attributes/text/attributes.md , it is very far from detailed. At most a sketch of wishes as many important details seem missing, including some key crucial technical ones. In particular, it doesn't address vernac execution at all, but only syntax. Also, the PR seems only to discuss attributes for declarations and inductives, I am very OK with them as long as they are done with the proper scoping.

In fact, my attribute record seems fairly orthogonal to the CEP.

> I would personally push for an attribute record that is extensible, even if it requires some form of dynamic type checking, and I've said so to @vbgl in private.

Well that system is just a map from `string` to `string` and while I am not opposed to that I do fail to see how having this can make the vernacular layer better, at least when we are talking about core functionality. 

I wish indeed we had a system to better classify commands, as to be able to say "this command supports locality, termination checking, and program mode". We could either use variants, GADT, or objects to achieve this.

Note that the CEP talks more about plugins, but here it seems clear to me that this guarded flag is better stored in the `VernacDefinition` constructor.

> I think you should just dismiss your "request for changes" to transform it into a simple "comment review".

In this case I am still feeling like strongly recommending to remove "VernacGuarded" case in favor of a command-specific entry, so I think I'll keep it for a while. I just want to highlight that I am not talking categorically here nor I am making any general statement about attributes, but I'd like people to reflect about it.

> The current proposal uses the attribute system internally but not on the user-side because there is no syntax for attributes yet.

So indeed it seems to me that the dual would be what was proposed and then would maybe help alleviate the parsing "problems". Parsing can be done in a more convenient way, however internally the new `Guarded` constructor doesn't make a lot of sense IMVHO.

File: vernac/vernacentries.ml
Comment (ejgallego) [reply vernac/vernacentries.ml]: I am not sure what do you mean, I do understand this mechanism as I basically started it and IIRC I suggested / discussed the current design with Maxime some time ago. Code like this seems quite fragile, even if I am the one to blame and it is not even usable by vernacs declared using the interp table. It seems clear to me that we should not foster this wrapping style but indeed declaration-only attributes should be properly scoped.

**Comment by gares** [comment]
> Simon explained that this new feature doesn't go as far as he would have liked because the check disabling only occurs at definition time. So if the fixpoint ended up in a proof it would have to be typechecked again and this would fail.

Right! Much better this way ;-)

**Comment by ejgallego** [comment]
@SimonBoulier personally I'd say that if you remove the `VernacGuarded` "attribute" and instead put this info in the `Definition` constructor itself that should be merged without too much problems.

Indeed it seems to me that quite a few design work on what "attributes" are and how are they gonna be implemented is needed, but IMHO that should be orthogonal to this PR.

**Comment by Zimmi48** [comment]
Except that this is precisely what @ppedrot opposed to, isn't it?

**Comment by ejgallego** [comment]
The way I understand it is that @ppedrot complains about the syntax, I am referring to the AST itself.

At least to me it is clear that the AST needs a bit more of design work so I think that what I suggest is a reasonable compromise in order not to get this PR stuck.

About the syntax, I have no particular strong opinion.

**Comment by ppedrot** [comment]
To be clear, I'm just opposed to the syntax extension. How hackish the implementation is can be frowned upon from an engineering point of view, but this is one order of magnitude less problematic as it does not appear directly to the user.

Quizz: without looking at the code, which permutation makes the syntax parsable in the current PR?
```
{Unguarded, Local, Polymorphic, Program} Fixpoint f (n : nat) := tt.
```

**Comment by mattam82** [comment]
I think Unguarded/General fxpoint could have even a larger scope than Program, and apply to Lemma, Inductive as well, ie any command where an unguarded fixpoint might be typechecked. In the current codebase there’s no parsing support for attributes so I didn’t see a problem with the implementation. When they’re available we’ll use them. I suppose there will be a mechanism to check the support of an attribute by a given command, @vbgl @maximedenes ?

**Comment by maximedenes** [comment]
> I suppose there will be a mechanism to check the support of an attribute by a given command, @vbgl @maximedenes ?

Yes, I believe that is the plan.

**Comment by Zimmi48** [comment]
> I think Unguarded/General fxpoint could have even a larger scope than Program, and apply to Lemma, Inductive as well, ie any command where an unguarded fixpoint might be typechecked.

For `Inductive`, do you mean you mean that `Unguarded` should give access to the feature that was implemented in #79? In that case, I wonder if `Unguarded` is the right generic terminology.

For `Lemma`, `Theorem`, etc, yes, this is what I suggested in https://github.com/coq/coq/pull/7651#pullrequestreview-124815588. However, I withdrew this in https://github.com/coq/coq/pull/7651#issuecomment-393805776 because I'm doubtful about the interest of using `Theorem` to define a (co-)fixpoint. I wonder if we shouldn't deprecate this feature instead. WDYT?

**Comment by SkySkimmer** [comment]
>For Inductive, do you mean you mean that Unguarded should give access to the feature that was implemented in #79? In that case, I wonder if Unguarded is the right generic terminology.

No, it's that you could write `Unguarded Inductive foo := bar : (fix f (x:unit) : empty := f x) -> foo.`

**Comment by nomeata** [comment]
This might be very useful to me! But I often work with `Definition` and local `fix`, and from the description it is unclear to me whether there is a way to mark a local `fix` as `Unguarded`. Will that be supported?

**Comment by Zimmi48** [comment]
It was discussed and this PR doesn't go as far because it doesn't change the AST of terms (we can't attach `Unguarded` to a term). However, I guess the options proposed in https://github.com/SimonBoulier/TypingFlags should do what you want. BTW @SimonBoulier what about first proposing to integrate these options to Coq. This was the approach recently followed by @jashug in #7703 to avoid the debates around attributes.

**Comment by SimonBoulier** [comment]
> what about first proposing to integrate these options to Coq

That's my plan! I do this as soon as my thesis is written  :-)

**Review by herbelin — COMMENTED** [review]
OK to me as far as I'm concerned.

**Review by gares — APPROVED** [review]
wrt the classifier

**Review by maximedenes — CHANGES_REQUESTED** [review]
This can now use attributes.

**Comment by maximedenes** [comment]
It seems to me that this PR could be rebased on top of attributes, now that we have them.

**Comment by Zimmi48** [comment]
I think this is simply waiting on @SimonBoulier finishing writing their thesis first.

**Comment by SimonBoulier** [comment]
Sorry I am not very available those days but I will do it.
What is the deadline for the 8.9? October, the 1st?

**Comment by Zimmi48** [comment]
October 1st is the date of the beta release so it will be too late for sure, but it is already quite late. If it is merged now, it goes in 8.10 by default unless developers convince the release manager how important it is to include this in 8.9.

**Comment by SimonBoulier** [comment]
Ok, it will be for 8.10 then.

**Comment by SimonBoulier** [comment]
Superseded by  #9004.

**Pull request closed without merging**
